# May 2024
Well I am creating this on May 12, 2024. So diary, I have quite a bit to catch up on.

## Start with a Look at the Phillies on Baseball Reference
Bohm is elite, hitting .343 (up 30% from his xBA) to produce a 170 OPS+. With RISP, he is hitting over .400 (+14%). <br>
Harper is producing despite not getting many hits. He has been unlucky (-17% from xBA) but has 9 dongs and wRC+ of 140. <br>
Turner was hot, then got hurt. <br>
Marsh has cooled off significantly. <br>
Castellanos is heating up but started off in a horrific slump and has a ways to go to catch up. <br>
Well, that is ok because Marsh started out pretty hot but is coming back to earth, he cannot face lefties (.162 xBA vs LHP). <br>
Schwarber slow to start but finding himself (pitches_per_plate_apps trending back to his Phillies mean of 7.3) <br>
JT doing his thing, swinging a consistent stick overall with runs of 10 games <br>
Stubbs stinks <br>
Stott slow start but finding his way <br>
Pache meh <br>
Rojas figuring it out? <br>
Edmundomania. Unserious baseball. .802 OPS. 2 OAA at SS. Slugs against LHP (.576 SLG vs LHP). <br>
Whit - slow start he will figure it out (regression to mean) <br>
Kody Clemens hits <br>
If I forgot anyone, then they must really stink.



### Pitchers
Wheels - Really battling his way with a dip in velo but he is effective. Lots of unearned runs :( <br>
Nola - shoving it, consistent, k/9 down but just elite run prevention. home_runs / plate_apps down 10& from career average. <br>
Suarez - lights out, using that cutter more. Lots of backwards Ks - how is he doing it? Sequencing? <br>
Sanchez - up and down but overall good results. Pitch mix (throw your fucking CH Cristian) <br>
Turnbull - Elite when in, slowing down, walks_per_plate_apps ticking up, still spinning that ST tho. <br>
Walker - Effective when on the mound, velo is down but he can throw that cutter and slider well. Less splitter () <br>
Hoffman - elite, throw that fucking SL <br>
Strahm - strikeout machine? What the fuck? Psycho <br>
Alvarado - saves, just overwhelming <br>
Dominguez - struggling <br>
Soto - -dealing but not getting results <br>
Yunior Marte - pitched well but ran into the ground. On the IL <br>
Orion - Not spinning it well in the early going but starting to figure it out? <br>
Ruiz - WHO? <br>
Luis F. Ortiz - hurt in early going <br>
Brogdon - pitched meaningful innings on the 2024 Philadelphia Phillies. <br>
Pinto - Joey bought his jersey <br>
Nelson - With Nick Nelson in AAA, who will smoke cigs in the locker room? <br>

## Run Functions

In [1]:
%run "Baseball Functions.ipynb"

## Imports and Data

In [2]:
import pandas as pd
import glob
import os
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

#%run "Baseball Functions.ipynb"
woba_weights = pd.read_csv('wOBA and FIP Constants.csv')

# Load All Non-Phillies Player Data
nphl = get_nphillies_data()

# Load Phillies Player Data With Function I wrote lol
phils_sc, pos, pps = get_phillies_data()

Length of nphl dataframe: 311816
Length of phils_sc dataframe: 422981


# SFG 53-56 (H)
I was in DC for this weekend. Well I went two of these four games lol.

## Game 1: Hicks vs Nola
Jordan Hicks is a starter now? Got sauced with the dentists, City Connect, offense battles.

## Game 2: vs Suarez
Another Suarez gem, ugly night, lots of offense.

## Game 3: Webb vs Walker
Sunday Night Baseball. Walker shoves. Harper 3-R Home Run. Phils are hot.

## Game 4: Black vs Wheeler
Wheeler gem. Way to battle and close out the four-game sweep you stud.

# TOR 57-58 (H)
Short two-game midweek stint for Phillies, split with the visiting Blue Jays.

## Game 1: Berrios vs Sanchez
Berrios battles but the offense is too good. Sanchez deals and Turnbull piggybacks to finish the last couple innings.

## Game 2: Bassitt vs Nola
The win streak ends, Nola is mortal. This team is not perfect.

# MIA 510-512 (A)
Weekend swing into Miami, bats were hot. Wheeler battles on Sunday but Phils cannot complete sweep.

## Game 1: Rogers vs Suarez
Dude, Ranger is legitimately pitching like a stud. Gold Glove and Cy Young. Venezuelan Greg Maddux.

## Game 2: Luzardo vs Walker
Walker and Luzardo duel. Phillies offense is just too good. Clicking.

## Game 3: Garrett vs Wheeler
Casty 3-R dong in the first but Wheeler has to grind and gets a bit unlucky. Offense forces extras but Fish pull it out in 10.

# NYM 513-514 (A)
Too bad I will be in NYC later this week. Let's keep the good vibes rolling. A split is pretty good here. Hopefully Cris Sanchez deals. Will Turnbull be available? He pitched Sunday but just one inning.

## Game 1: Manaea vs Sanchez
How do the Phillies match up against Manaea or pitchers like him? <br>
How do the Mets match up against Sanchez or pitcher likes him? <br>

In [3]:
# Read in Manaea Data
manaea = pd.read_csv('manaea.csv')

# Define Manaea vs LHB
manaea_lhb_career = lhb_pitch_mix(manaea)
manaea_lhb_2024 = lhb_pitch_mix(manaea[manaea.game_year == 2024])

# Define Manaea vs RHB
manaea_rhb_career = rhb_pitch_mix(manaea)
manaea_rhb_2024 = rhb_pitch_mix(manaea[manaea.game_year == 2024])

# Define Manaea ID
manaea.groupby('player_name',as_index=False).agg({'pitcher' : 'max'})
manaea_id = 640455

# Define Phillies vs Manaea
manaea_vs_phils = pos[pos.pitcher == manaea_id]

# Define Non-Phillies vs Manaea
manaea_vs_nphils = nphl[nphl.pitcher == manaea_id]

In [4]:
manaea.groupby('game_year',as_index=False).agg({'des' : 'count'})

game_year   des
0       2016  2169
1       2017  2688
2       2018  2371
3       2019   444
4       2020   812
5       2021  2980
6       2022  2482
7       2023  2052
8       2024   636

In [5]:
df = manaea_vs_phils
results('game_year')

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2017      101          25   19     6  0.0      2           4  0.261   
1       2022       34           9    5     4  1.0      2           2  0.571   
2       2023       84          18   11     4  3.0      2           5  0.267   
3       2024      100          23   16     4  0.0      1           6  0.182   

     obp    slg    ops   woba  
0  0.320  0.304  0.624  0.280  
1  0.667  1.286  1.952  0.762  
2  0.333  1.000  1.333  0.499  
3  0.217  0.227  0.445  0.202

Sean Manaea has pitched three times against the Phillies in his career.
- In 2017, he shoved. He also faced a bunch of idiots that day.
- In the 2022 NLCS, he got hit around. He gave up a 1.952 OPS.
- Last year, he gave up 3 Home Runs in 18 plate appearances. ONly four hits, but he got rocked.

In [6]:
manaea_pos_results = results('player_name').sort_values(by='plate_apps',ascending=False)
manaea_pos_results[manaea_pos_results.player_name.isin(['Stott, Bryson'
                                                        ,'Bohm, Alec'
                                                        ,'Harper, Bryce'
                                                        ,'Castellanos, Nick'
                                                        ,'Marsh, Brandon'
                                                        ,'Sosa, Edmundo'
                                                        ,'Rojas, Johan'
                                                       ]
                                                      )
                  ]

player_name  pitches  plate_apps  bip  hits  hrs  walks  strikeouts  \
19      Stott, Bryson       39           7  6.0   3.0  1.0    1.0         0.0   
2          Bohm, Alec       18           6  4.0   1.0  1.0    0.0         2.0   
7       Harper, Bryce       27           6  5.0   2.0  0.0    0.0         1.0   
3   Castellanos, Nick       25           5  2.0   1.0  0.0    0.0         3.0   
11     Marsh, Brandon       18           4  1.0   0.0  0.0    1.0         2.0   
18      Sosa, Edmundo       11           4  3.0   3.0  1.0    0.0         1.0   
14       Rojas, Johan       16           3  2.0   0.0  0.0    0.0         1.0   

       ba    obp   slg    ops   woba  
19  0.500  0.571  1.00  1.571  0.640  
2   0.200  0.167  0.80  0.967  0.334  
7   0.333  0.333  0.50  0.833  0.359  
3   0.200  0.200  0.40  0.600  0.252  
11  0.000  0.250  0.00  0.250  0.174  
18  0.750  0.750  2.25  3.000  1.212  
14  0.000  0.000  0.00  0.000  0.000

KS Projected Lineup
1. Stott 4 - Has hit pretty well against Manaea in his career and hits these sorts of lefties pretty well.
2. Bohm 5 - Hits well against Manaea, is hot, and hits guys like Manaea well.
3. Harper 3 - Ok against Manaea in his career.
4. Castellanos 0 - Hot and can drive the ball against Manaea and guys like him in his career.
5. Marsh 7 - Punched out twice by Manaea, neither AB looked comfortable. Decent against guys like Manaea.
6. Sosa 6 - Has two ABs against Manaea. HR and 3B. Both well struck to CF. Hits guys like Manaea well. He is hot.
7. Pache 9 - Hits lefties like Manaea well.
8. Stubbs 2 - Struggles against guys like Manaea but JT does not have a ton of success and is still hurt. Probably my hottest take.
9. Rojas 8 - Only a couple PAs against guys like Manaea.

After review, Whit for Marsh seems like a must. Maybe swap out Stubbs if Sanchez pitches significantly better to JT.
1. Whit 7 - Rakes against Manaea and guys like him in his career
2. JT 2 - I don't think you can keep JT out of the lineup unless Stubbs rakes against archetype.
3. Harper 3 - 1/3 in his career against Manaea. Bashes against archetype.
4. Bohm 5 - Likes Manaea, hit a homer against him. Hits well of Manaea-esque
5. Castellanos 0 - Has hit Manaea well in his career.
6. Stott 4 - Hits pretty well against Manaea in his career, well he has some interesting splits against these sorts of lefty pitches.
7. Sosa 6 - Has two ABs against Manaea. HR and 3B. Both well struck to CF. He is hot. Solid against archetype.
8. Pache 9 - Hot and likes these sorts of lefties
9. Rojas 8 - Maybe too many righties here but well I guess there are only two LHB in the whole lineup. Meh against archetype.

In [7]:
df = manaea_vs_nphils
manaea_nphl_results = results('player_name').sort_values(by='plate_apps',ascending=False)
manaea_nphl_results[manaea_nphl_results.player_name.isin(['Merrifield, Whit', 'Castellanos, Nick'])]

player_name  pitches  plate_apps  bip  hits  hrs  walks  strikeouts  \
5   Merrifield, Whit       37           9  6.0   4.0  0.0    1.0         2.0   
0  Castellanos, Nick       14           6  6.0   4.0  2.0    0.0         0.0   

      ba    obp    slg    ops   woba  
5  0.500  0.556  0.750  1.306  0.549  
0  0.667  0.667  1.667  2.333  0.967

In [8]:
# Manaea Pitch Profile in Career
manaea.groupby('pitch_type',as_index=False).agg({'release_speed' : ['min','max', 'mean', 'std']
                                                 ,'release_spin_rate' : ['min','max', 'mean', 'std']
                                                }).round(1)

pitch_type release_speed                  release_spin_rate                  \
                       min   max  mean  std               min     max    mean   
0         CH          77.2  92.7  84.1  2.1             592.0  3365.0  1310.5   
1         CU          75.5  81.0  78.1  2.3            1832.0  2102.0  1999.9   
2         FC          81.0  92.5  87.6  2.2             993.0  3001.0  2009.2   
3         FF          86.3  97.2  92.5  1.8            1040.0  2942.0  2012.0   
4         SI          85.6  96.2  91.3  1.6            1597.0  2987.0  1980.4   
5         SL          73.0  91.1  81.0  2.9             648.0  3458.0  2001.7   
6         ST          72.1  83.9  78.9  2.1            1561.0  3064.0  2119.3   

          
     std  
0  252.9  
1   85.4  
2  171.0  
3   96.6  
4  100.2  
5  150.9  
6  144.0

In [9]:
manaea_lhb_2024

pitch_type       pitch_name  count  release_speed  release_spin_rate  \
3         SI           Sinker     39      92.697436        2014.512821   
4         ST          Sweeper     24      78.366667        2192.541667   
2         FF  4-Seam Fastball      8      92.087500        2005.375000   
0         CH         Changeup      5      87.140000        1475.400000   
1         FC           Cutter      1      89.400000        1999.000000   

        zone  usage  
3   6.897436   50.6  
4   8.416667   31.2  
2   7.000000   10.4  
0  13.400000    6.5  
1   4.000000    1.3

In [10]:
df = pos[(pos.p_throws == 'L')
    &(pos.pitch_type == 'SI')
    &(pos.release_speed >= 91.3-1.6)
    &(pos.release_speed <= 91.3+1.6)
    &(pos.release_spin_rate >= 1980.4-100.2)
    &(pos.release_spin_rate <= 1980.4+100.2)
    &(pos.player_name.isin(['Stott, Bryson', 'Harper, Bryce', 'Stubbs, Garrett']))
   ]
results('player_name')

player_name  pitches  plate_apps  bip  hits  hrs  walks  strikeouts  \
0    Harper, Bryce      103          35   27   7.0  0.0    4.0           3   
1    Stott, Bryson       57          11    9   3.0  0.0    0.0           2   
2  Stubbs, Garrett       16           5    3   0.0  0.0    1.0           1   

      ba    obp    slg    ops   woba  
0  0.241  0.343  0.310  0.653  0.296  
1  0.273  0.273  0.364  0.636  0.276  
2  0.000  0.200  0.000  0.200  0.138

In [11]:
df = pos[(pos.p_throws == 'L')
    &(pos.pitch_type == 'ST')
    &(pos.release_speed >= 78.9-2.1)
    &(pos.release_speed <= 78.9+2.1)
    #&(pos.release_spin_rate >= 2119.3-144)
    #&(pos.release_spin_rate <= 2119.3+144)         
    &(pos.player_name.isin(['Stott, Bryson', 'Harper, Bryce', 'Stubbs, Garrett']))
   ]
results('player_name')

player_name  pitches  plate_apps  bip  hits  hrs  walks  strikeouts  \
0    Harper, Bryce       49          12    7     3  1.0    2.0         3.0   
1    Stott, Bryson       26           6    2     1  0.0    1.0         3.0   
2  Stubbs, Garrett        5           1    1     1  0.0    0.0         0.0   

    ba    obp  slg    ops   woba  
0  0.3  0.417  0.6  1.017  0.430  
1  0.2  0.333  0.2  0.533  0.265  
2  1.0  1.000  1.0  2.000  0.894

In [12]:
manaea_rhb_2024

pitch_type       pitch_name  count  release_speed  release_spin_rate  \
3         SI           Sinker    176      92.387500        2034.738636   
0         CH         Changeup    108      85.306481        1476.629630   
2         FF  4-Seam Fastball    102      91.947059        2004.362745   
1         FC           Cutter     84      88.390476        2004.583333   
5         ST          Sweeper     72      77.195833        2181.597222   
4         SL           Slider     17      85.629412        1968.176471   

        zone  usage  
3   7.948864   31.5  
0  10.731481   19.3  
2   8.196078   18.2  
1   8.428571   15.0  
5  10.236111   12.9  
4   9.823529    3.0

In [13]:
df = pos[(pos.p_throws == 'L')
    &(pos.pitch_type == 'SI')
    &(pos.release_speed >= 91.3-1.6)
    &(pos.release_speed <= 91.3+1.6)
    #&(pos.release_spin_rate >= 1980.4-100.2)
    #&(pos.release_spin_rate <= 1980.4+100.2)
    &(pos.player_name.isin(['Bohm, Alec'
                            ,'Realmuto, J.T.'
                            ,'Castellanos, Nick'
                            ,'Sosa, Edmundo'
                            ,'Pache, Cristian'
                            ,'Rojas, Johan'
                            ,'Merrifield, Whit'
                           ]
                          )
     )
   ]
results('player_name')

player_name  pitches  plate_apps  bip  hits  hrs  walks  strikeouts  \
0         Bohm, Alec      149          36   25     5  1.0    3.0         8.0   
1  Castellanos, Nick       73          22   14     4  1.0    2.0         6.0   
2   Merrifield, Whit       16           7    6     2  0.0    0.0         1.0   
3    Pache, Cristian       28           7    4     1  0.0    3.0         0.0   
4     Realmuto, J.T.      197          53   40    17  0.0    6.0         7.0   
5       Rojas, Johan       31           4    3     1  0.0    0.0         1.0   
6      Sosa, Edmundo       48          11    7     2  0.0    0.0         2.0   

      ba    obp    slg    ops   woba  
0  0.152  0.222  0.273  0.495  0.222  
1  0.200  0.273  0.400  0.673  0.295  
2  0.286  0.286  0.286  0.571  0.255  
3  0.250  0.571  0.250  0.821  0.424  
4  0.362  0.434  0.511  0.945  0.410  
5  0.250  0.250  0.250  0.500  0.221  
6  0.222  0.364  0.556  0.919  0.395

In [14]:
pps[(pps.game_year == 2024) & (pps.p_throws == 'L')].groupby('player_name',as_index=False).agg({'pitcher' : 'max'
                                                                                                ,'des' : 'count'
                                                                                               }).sort_values(by='des',ascending=False)
alvarado_id = 621237

In [15]:
df = pps[(pps.pitcher == alvarado_id)
    &(pps.home_team == 'NYM')
   ]
results('stand')

stand  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba    obp  \
0     L       57          16   11     1  0.0      1           3  0.071  0.188   
1     R       71          18    8     4  0.0      1           9  0.235  0.278   

     slg    ops   woba  
0  0.071  0.259  0.144  
1  0.294  0.572  0.254

## Game 2: Butto vs Nola
Nola owns the Mets? <br>
Jose BUTTo vs the Phillies <br>

In [16]:
butto = pd.read_csv('butto.csv')
butto.groupby('game_year',as_index=False).agg({'des' : 'count'})

game_year  des
0       2022   98
1       2023  733
2       2024  554

# NYM 515-516 (H)
## Game 1: vs Suarez
## Game 2: vs Walker

# WAS 517-519 (H)
## Game 1: Irwin vs Wheeler

## Game 2: Gore vs Sanchez

In [17]:
pps[(pps.game_year == 2024) & (pps.p_throws == 'L')].groupby('player_name',as_index=False).agg({'des' : 'count'
                                                                                                ,'pitcher' : 'max'
                                                                                               }).sort_values(by='des',ascending=False)
sanchez_id = 650911

sanchez24 = pps[(pps.game_year == 2024) & (pps.pitcher == sanchez_id)]
sanchez = pps[pps.pitcher == sanchez_id]
sanchezWSH = sanchez[(sanchez.home_team == 'WSH') | (sanchez.away_team == 'WSH')]

In [18]:
pps[pps.game_year == pps.game_year.max()].groupby('home_team',as_index=False).agg({'game_pk' : 'nunique'}
                                                                                 ).sort_values(by='game_pk',ascending=False)

home_team  game_pk
5       PHI       30
0       CIN        4
1       COL        3
2       LAA        3
3       MIA        3
6        SD        3
7       STL        3
8       WSH        3
4       NYM        2

In [19]:
df = sanchezWSH = sanchezWAS
results('game_date')

NameError: name 'sanchezWAS' is not defined

## Cris Sanchez Start vs WSH on 4/7/2024
First, my notes from this journal. <br>
N/a <br>

Wonderful.


In [20]:
pps[pps.game_date == '2024-04-07'
   ].groupby('game_pk', as_index=False
            ).agg({'post_home_score' : 'max'
                   ,'home_team' : 'max'
                   ,'post_away_score' : 'max'
                   ,'away_team' : 'max'
                  })

game_pk  post_home_score home_team  post_away_score away_team
0   744871                3       WSH                2       PHI

In [21]:
def game_summary(df):
    x = df.groupby('game_date', as_index=False
              ).agg({'game_pk' : 'max'
                     ,'home_team' : 'max'
                     ,'post_home_score' : 'max'
                     ,'away_team' : 'max'
                     ,'post_away_score' : 'max'
                     ,'estimated_ba_using_speedangle' : 'mean'
                     ,'estimated_woba_using_speedangle' : 'mean'
                     ,'des' : 'count'
                     ,'pitcher' : 'nunique'
                     ,'inning' : 'nunique'
                     ,'at_bat_number' : 'nunique'
                    }).round(3)
    y = results('home_team')
    gs = x.merge(y, left_on = 'home_team', right_on = 'home_team', suffixes = ('','_y'))
    return gs

game_summary(df)

game_date  game_pk home_team  post_home_score away_team  post_away_score  \
0  2021-04-13   632189       NYM                2       PHI                2   
1  2021-06-25   633545       NYM                1       PHI                1   
2  2021-06-27   633504       NYM                1       PHI                4   
3  2021-09-17   632467       NYM                1       PHI                2   
4  2021-09-18   632447       NYM                1       PHI                2   
5  2022-04-30   662591       NYM                1       PHI                0   
6  2022-05-01   662579       NYM                6       PHI                3   
7  2022-08-12   662434       NYM                1       PHI                1   
8  2023-09-30   716393       NYM               11       PHI                4   
9  2024-05-13   745826       NYM                4       PHI                5   

   estimated_ba_using_speedangle  estimated_woba_using_speedangle  des  \
0                          0.094                            0.115   12   
1                          0.382                            0.368   31   
2                          0.539                            0.506   20   
3                          0.280                            0.260   13   
4                            NaN                              NaN    3   
5                            NaN                              NaN    7   
6                          0.337                            0.316   12   
7                          0.328                            0.306    8   
8                          0.034                            0.031   13   
9                          0.323                            0.296    9   

   pitcher  ...  bip  hits  hrs  walks  strikeouts     ba    obp    slg  \
0        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
1        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
2        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
3        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
4        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
5        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
6        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
7        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
8        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   
9        1  ...   19     5  0.0      2          12  0.161  0.235  0.194   

     ops   woba  
0  0.429  0.202  
1  0.429  0.202  
2  0.429  0.202  
3  0.429  0.202  
4  0.429  0.202  
5  0.429  0.202  
6  0.429  0.202  
7  0.429  0.202  
8  0.429  0.202  
9  0.429  0.202  

[10 rows x 24 columns]

In [22]:
df = game_summary(df)
df.columns.tolist()

['game_date',
 'game_pk',
 'home_team',
 'post_home_score',
 'away_team',
 'post_away_score',
 'estimated_ba_using_speedangle',
 'estimated_woba_using_speedangle',
 'des',
 'pitcher',
 'inning',
 'at_bat_number',
 'pitches',
 'plate_apps',
 'bip',
 'hits',
 'hrs',
 'walks',
 'strikeouts',
 'ba',
 'obp',
 'slg',
 'ops',
 'woba']

In [23]:
# nick castellanos by temperature
casty_id = 592206
casty = pd.concat([pos[pos.batter == casty_id], nphl[nphl.player_name == 'Castellanos, Nick']])

In [24]:
philly_weather = pd.read_csv('C://Users//Kellen//Downloads//Philly Weather 2015 to 20230707.csv')
philly_weather

DATE_VALUE  MAX_TEMPERATURE  RAINFALL_AMT
0     2015-01-01            36.82      0.000000
1     2015-01-02            43.02      0.000000
2     2015-01-03            43.41      0.085521
3     2015-01-04            48.16      0.930987
4     2015-01-05            54.21      0.136869
...          ...              ...           ...
3119  2023-07-16            86.60      0.076772
3120  2023-07-17            92.40      0.059055
3121  2023-07-18            95.50      0.000000
3122  2023-07-19            91.50      0.000000
3123  2023-07-20            89.70      0.000000

[3124 rows x 3 columns]

In [25]:
casty_cbp = casty[casty.home_team == 'PHI']
df = casty_cbp.merge(philly_weather, left_on = 'game_date', right_on = 'DATE_VALUE', how = 'left', suffixes = ('','_weather'))
results('MAX_TEMPERATURE')

MAX_TEMPERATURE  pitches  plate_apps  bip  hits  hrs  walks  strikeouts  \
0              54.01       14           4  1.0   0.0  0.0    2.0         1.0   
1              54.61       10           4  3.0   0.0  0.0    0.0         1.0   
2              54.82       20           4  2.0   1.0  0.0    1.0         1.0   
3              55.54       15           4  2.0   1.0  0.0    0.0         2.0   
4              56.01       27           8  5.0   1.0  0.0    0.0         3.0   
..               ...      ...         ...  ...   ...  ...    ...         ...   
109            96.19       15           4  4.0   2.0  0.0    0.0         0.0   
110            97.05       16           4  2.0   0.0  0.0    1.0         0.0   
111            97.23       13           5  4.0   0.0  0.0    0.0         1.0   
112            97.81       18           4  3.0   0.0  0.0    0.0         1.0   
113            98.60       16           4  2.0   1.0  0.0    0.0         2.0   

        ba    obp    slg    ops   woba  
0    0.000  0.500  0.000  0.500  0.348  
1    0.000  0.000  0.000  0.000  0.000  
2    0.333  0.500  0.333  0.833  0.393  
3    0.250  0.250  0.500  0.750  0.311  
4    0.125  0.125  0.125  0.250  0.110  
..     ...    ...    ...    ...    ...  
109  0.500  0.500  0.500  1.000  0.442  
110  0.000  0.250  0.000  0.250  0.172  
111  0.000  0.000  0.000  0.000  0.000  
112  0.000  0.000  0.000  0.000  0.000  
113  0.250  0.250  0.250  0.500  0.221  

[114 rows x 13 columns]

In [26]:
df

pitch_type   game_date  release_speed  release_pos_x  release_pos_z  \
0            CU  2022-08-28           76.6          -0.78           6.11   
1            SL  2022-08-28           81.8          -0.83           5.90   
2            SL  2022-08-28           82.4          -0.79           5.92   
3            SL  2022-08-28           83.2          -0.79           6.05   
4            FF  2022-08-28           92.9          -0.99           5.94   
...         ...         ...            ...            ...            ...   
2937         SL  2019-04-30           88.3          -2.43           5.54   
2938         FF  2019-04-30           94.4          -2.44           5.73   
2939         SI  2019-04-30           98.0          -2.67           5.36   
2940         FF  2019-04-30           95.0          -2.54           5.70   
2941         FF  2019-04-30           97.0          -2.55           5.41   

            player_name  batter  pitcher     events      description  ...  \
0     Castellanos, Nick  592206   672710  field_out    hit_into_play  ...   
1     Castellanos, Nick  592206   672710  strikeout  swinging_strike  ...   
2     Castellanos, Nick  592206   672710        NaN             ball  ...   
3     Castellanos, Nick  592206   672710        NaN  swinging_strike  ...   
4     Castellanos, Nick  592206   672710        NaN             ball  ...   
...                 ...     ...      ...        ...              ...  ...   
2937  Castellanos, Nick  592206   622554  strikeout  swinging_strike  ...   
2938  Castellanos, Nick  592206   592826        NaN  swinging_strike  ...   
2939  Castellanos, Nick  592206   622554        NaN             foul  ...   
2940  Castellanos, Nick  592206   592826        NaN             ball  ...   
2941  Castellanos, Nick  592206   622554        NaN             foul  ...   

        wHR  runSB  runCS   R/PA     R/W   cFIP source  DATE_VALUE  \
0     2.072    0.2 -0.397  0.114   9.524  3.112    NaN  2022-08-28   
1     2.072    0.2 -0.397  0.114   9.524  3.112    NaN  2022-08-28   
2     2.072    0.2 -0.397  0.114   9.524  3.112    NaN  2022-08-28   
3     2.072    0.2 -0.397  0.114   9.524  3.112    NaN  2022-08-28   
4     2.072    0.2 -0.397  0.114   9.524  3.112    NaN  2022-08-28   
...     ...    ...    ...    ...     ...    ...    ...         ...   
2937  1.940    0.2 -0.435  0.126  10.296  3.214      n  2019-04-30   
2938  1.940    0.2 -0.435  0.126  10.296  3.214      n  2019-04-30   
2939  1.940    0.2 -0.435  0.126  10.296  3.214      n  2019-04-30   
2940  1.940    0.2 -0.435  0.126  10.296  3.214      n  2019-04-30   
2941  1.940    0.2 -0.435  0.126  10.296  3.214      n  2019-04-30   

     MAX_TEMPERATURE RAINFALL_AMT  
0              89.20     0.000000  
1              89.20     0.000000  
2              89.20     0.000000  
3              89.20     0.000000  
4              89.20     0.000000  
...              ...          ...  
2937           68.97     0.068898  
2938           68.97     0.068898  
2939           68.97     0.068898  
2940           68.97     0.068898  
2941           68.97     0.068898  

[2942 rows x 112 columns]

In [27]:
x = df.MAX_TEMPERATURE
y = df.ops
title = 'Nick Castellanos OPS by Max Temperature of the Day, Colored by Year'
subtitle = ''
graph(x,y,title,color='#003469')

AttributeError: 'DataFrame' object has no attribute 'ops'

## Game 3: Williams vs Nola

Trevor Williams is really good? The Phillies got him out of the game in the 5th inning. <br>
Maybe Trevor Williams could be a Phillies option? <br>
Spencer Turnbull as a trade deadline piece -- You never get rid of good pitching depth. <br>
Kody Clemens ABs + Defense <br>
Harper cruises into second with a 2-R double. <br>
Bohm longest HR of career? No, he has 20 career home runs that have traveled further than that. 40% of the home runs Alec Bohm has hit<br> in his career have traveled further than his 8th inning three-run dong off of Tanner Rainey to make it 11 to 5 Phillies. Pimp job.

In [ ]:
bohm_id = pos.groupby('player_name',as_index=False).agg({'batter' : 'max', 'des' : 'count'}).sort_values(by='des',ascending=False)

In [ ]:
bohm_id = 664761

In [28]:
pos[(pos.batter == bohm_id)
    &(pos.events == 'home_run')
   ].groupby('player_name',as_index=False).agg({'hit_distance_sc' : 'max'})

NameError: name 'bohm_id' is not defined

In [29]:
pos[(pos.batter == bohm_id)
    &(pos.events == 'home_run')
    #&(pos.hit_distance_sc == 446)
   ][['home_team','game_date','hit_distance_sc','launch_angle','launch_speed']].sort_values(by='hit_distance_sc',ascending=False
                                                                                           ).reset_index(drop=True
                                                                                                        )

NameError: name 'bohm_id' is not defined

In [30]:
len(pos[(pos.batter == bohm_id)&(pos.events == 'home_run')])

NameError: name 'bohm_id' is not defined

In [31]:
20/51

0.39215686274509803

In [32]:
pct_share = 40
print('{}% of the home runs hit by {} in his career have traveled further than his {} home run against the {} that went {} ft'.format(str(pct_share)
              ,pos[pos.batter == bohm_id].player_name.max() # player_name
              ,pos[(pos.batter == bohm_id)&(pos.game_date == '2024-05-19')].game_date.max()#'May 19th'#game_date
              ,pos[(pos.game_date == '2024-05-19')].away_team.max()#WSH'#home_team or away_team
              ,pos[(pos.batter == bohm_id)&(pos.events == 'home_run')&(pos.game_date == '2024-05-19')].hit_distance_sc.max()#'446'
             )
     )

NameError: name 'bohm_id' is not defined

In [33]:
pos[(pos.batter == bohm_id) & (pos.events == 'home_run') #& (pos.game_date == pos.game_date.max())
   ].hit_distance_sc.rank(0,'dense'
                          #,na_option = 'bottom'
                          ,ascending=False
                         )

NameError: name 'bohm_id' is not defined

# TEX 521-523 (H)
## Game 1: vs Walker
## Game 2: vs Suarez
## Game 3: vs Wheeler

## COL 524-526 (H)
## Game 1: vs Nola
## Game 2: vs Sanchez
## Game 3: vs Suarez

# SFG 527-30 (A)
## Game 1: Snell vs Walker
## Game 2: Howard vs Wheeler
## Game 3: vs Suarez

## STL 531-62 (H)

## Game 1: Mikolas vs Nola

## Game 2: Gray vs Suarez

## Game 3: Lynn vs Walker

# Just a Random Fucking Thing - Phillies Extra Innings Results

In [34]:
df = pos[pos.inning > 9]
results('game_year').sort_values(by='plate_apps',ascending=False)

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
4       2019      569         143   96    21  4.0     15          30  0.168   
2       2017      512         137   97    37  1.0      7          33  0.291   
3       2018      540         135   82    22  3.0     21          32  0.198   
0       2015      480         120   87    25  3.0      8          23  0.236   
1       2016      343          95   62    22  1.0      6          19  0.282   
8       2023      353          88   60    18  1.0      8          18  0.250   
7       2022      281          67   46    14  2.0      5          15  0.241   
6       2021      200          56   43    20  2.0      1          11  0.385   
9       2024      149          37   22     5  0.0      5           9  0.179   
5       2020       85          20   16     6  2.0      1           3  0.375   

     obp    slg    ops   woba  
4  0.266  0.336  0.602  0.261  
2  0.321  0.354  0.675  0.293  
3  0.319  0.333  0.652  0.292  
0  0.283  0.358  0.642  0.277  
1  0.326  0.423  0.749  0.312  
8  0.318  0.333  0.652  0.285  
7  0.299  0.362  0.661  0.288  
6  0.375  0.596  0.971  0.398  
9  0.270  0.214  0.485  0.225  
5  0.350  0.750  1.100  0.409

# Phillies Strength of Schedule

In [35]:
po24 = pos[pos.game_year == 2024]
#standings = pd.read_csv('standings.csv')
#po24.merge(standings,left_on = 'home_team', right_on = '', suffixes = ('','_y'))

In [36]:
x = po24.groupby('game_date',as_index=False).agg({'launch_speed' : 'max'
                                              ,'estimated_ba_using_speedangle' : 'mean'
                                             }).sort_values(by='launch_speed',ascending=False).reset_index(drop=True)

In [37]:
df = pos.merge(x, left_on = 'launch_speed', right_on = 'launch_speed', suffixes = ('','_x'))
df.groupby('game_year',as_index=False).agg({'des' : 'count', 'game_pk' : 'nunique'})

game_year  des  game_pk
0       2015  152       87
1       2016  200       95
2       2017  241      104
3       2018  215      106
4       2019  278      115
5       2020   76       37
6       2021  269      110
7       2022  332      143
8       2023  339      128
9       2024  136       54

In [38]:
# These two lines prove to me that launch_speed is a good thing
# No wait what did I actually test here? I joined from x on the launch_speeds that have been the maximum launch speed in the 48 game season
x.agg({'launch_speed' : ['min','max', 'mean', 'std']}).round(3)

x.sort_values(by='launch_speed')

game_date  launch_speed  estimated_ba_using_speedangle
53  2024-05-07         103.4                       0.427348
52  2024-05-11         103.5                       0.402885
51  2024-05-04         104.0                       0.337844
50  2024-04-16         105.7                       0.309409
49  2024-04-01         106.1                       0.247040
48  2024-05-23         106.4                       0.396552
47  2024-05-08         106.4                       0.303370
46  2024-05-14         106.6                       0.308808
45  2024-05-26         106.7                       0.282292
44  2024-05-03         107.2                       0.371773
43  2024-05-22         107.3                       0.369458
42  2024-04-19         107.3                       0.260652
41  2024-04-24         107.4                       0.356536
40  2024-05-15         107.4                       0.350613
39  2024-05-12         107.7                       0.375440
37  2024-04-15         107.8                       0.243600
38  2024-04-05         107.8                       0.316600
36  2024-04-06         107.9                       0.341038
35  2024-04-27         107.9                       0.353680
34  2024-04-28         108.0                       0.385464
33  2024-04-02         108.3                       0.337966
32  2024-05-18         108.4                       0.294903
31  2024-04-14         108.5                       0.306778
30  2024-05-19         108.6                       0.402862
29  2024-05-16         108.9                       0.276969
28  2024-04-23         108.9                       0.233560
27  2024-04-12         108.9                       0.336231
26  2024-04-07         109.2                       0.332316
24  2024-05-01         109.4                       0.483533
25  2024-04-21         109.4                       0.319036
23  2024-04-29         109.6                       0.347963
22  2024-03-30         109.7                       0.330786
21  2024-05-24         109.8                       0.288438
20  2024-04-13         109.8                       0.308320
19  2024-04-08         110.1                       0.291735
18  2024-04-11         110.3                       0.336952
17  2024-05-10         110.4                       0.391206
16  2024-04-26         110.8                       0.385419
15  2024-04-17         110.8                       0.331040
14  2024-05-05         111.0                       0.418778
13  2024-05-13         111.0                       0.307296
12  2024-03-29         111.1                       0.305059
11  2024-05-06         111.5                       0.336852
10  2024-04-30         111.6                       0.357731
9   2024-04-25         112.0                       0.346833
8   2024-05-21         112.5                       0.475348
7   2024-05-25         112.7                       0.370909
6   2024-04-20         112.8                       0.287833
5   2024-04-03         113.0                       0.284318
4   2024-04-22         113.3                       0.290966
3   2024-03-31         114.4                       0.468318
2   2024-05-17         114.6                       0.194760
1   2024-04-09         115.1                       0.298947
0   2024-04-10         115.2                       0.275409

In [39]:
results('game_year')

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2015      152         134  134    94   21    0.0         0.0  0.701   
1       2016      200         181  181   117   37    0.0         0.0  0.646   
2       2017      241         211  211   136   37    0.0         0.0  0.648   
3       2018      215         200  200   152   55    0.0         0.0  0.760   
4       2019      278         252  252   159   49    0.0         0.0  0.633   
5       2020       76          61   61    41   15    0.0         0.0  0.672   
6       2021      269         235  235   173   59    0.0         0.0  0.736   
7       2022      332         271  271   191   80    0.0         0.0  0.705   
8       2023      339         288  288   207   73    0.0         0.0  0.719   
9       2024      136         115  115    75   20    0.0         0.0  0.652   

     obp    slg    ops   woba  
0  0.701  1.425  2.127  0.898  
1  0.646  1.392  2.039  0.848  
2  0.645  1.424  2.068  0.846  
3  0.760  1.790  2.550  1.060  
4  0.631  1.402  2.033  0.820  
5  0.672  1.475  2.148  0.886  
6  0.736  1.711  2.447  1.010  
7  0.705  1.753  2.458  1.035  
8  0.719  1.677  2.396  0.990  
9  0.652  1.322  1.974  0.848

In [40]:
df.columns[len(df.columns)-10:len(df.columns)]
gh = df[df.game_date == df.game_date_x].groupby('game_pk',as_index=False).agg({'des' : 'count'})
jk = pos[pos.game_pk == gh[gh.des>1].game_pk.max()]

In [41]:
jk[jk.launch_speed.isin(x.launch_speed.unique().tolist())
  ][['launch_speed'
     ,'game_date'
     ,'estimated_ba_using_speedangle'
     ,'player_name'
     ,'batter'
     ,'events'
     ,'launch_angle'
    ]]

launch_speed   game_date  estimated_ba_using_speedangle  \
206352         110.4  2024-05-10                          0.923   
206360         106.6  2024-05-10                          0.257   
206465         110.4  2024-05-10                          0.780   

            player_name  batter     events  launch_angle  
206352     Rojas, Johan  679032     single          11.0  
206360  Schwarber, Kyle  656941  field_out          45.0  
206465    Harper, Bryce  547180     double          18.0

In [42]:
dh = df.groupby('game_date',as_index=False).agg({'game_pk' : 'nunique'})
dh = dh[dh.game_pk > 1]
dh

game_date  game_pk
74   2015-09-11        2
204  2017-05-14        2
375  2018-09-11        2
528  2020-09-08        2
529  2020-09-11        2
531  2020-09-18        2
607  2021-07-29        2
671  2022-05-08        2
698  2022-06-17        2
737  2022-08-20        2
800  2023-04-18        2
885  2023-09-11        2

In [43]:
df = df[df.game_year == 2024]
results('events')

events  pitches  plate_apps  bip  hits   hrs  walks  \
0                     double       17          17   17  17.0   0.0    0.0   
1                double_play        1           1    1   0.0   0.0    0.0   
2                  field_out       31          31   31   0.0   0.0    0.0   
3            fielders_choice        1           1    1   0.0   0.0    0.0   
4                  force_out        3           3    3   0.0   0.0    0.0   
5  grounded_into_double_play        4           4    4   0.0   0.0    0.0   
6                   home_run       20          20   20  20.0  20.0    0.0   
7                     single       38          38   38  38.0   0.0    0.0   

   strikeouts   ba  obp  slg  ops   woba  
0         0.0  1.0  1.0  2.0  3.0  1.276  
1         0.0  0.0  0.0  0.0  0.0  0.000  
2         0.0  0.0  0.0  0.0  0.0  0.000  
3         0.0  0.0  0.0  0.0  0.0  0.000  
4         0.0  0.0  0.0  0.0  0.0  0.000  
5         0.0  0.0  0.0  0.0  0.0  0.000  
6         0.0  1.0  1.0  4.0  5.0  2.093  
7         0.0  1.0  1.0  1.0  2.0  0.894

In [44]:
results('events')

events  pitches  plate_apps  bip  hits   hrs  walks  \
0                     double       17          17   17  17.0   0.0    0.0   
1                double_play        1           1    1   0.0   0.0    0.0   
2                  field_out       31          31   31   0.0   0.0    0.0   
3            fielders_choice        1           1    1   0.0   0.0    0.0   
4                  force_out        3           3    3   0.0   0.0    0.0   
5  grounded_into_double_play        4           4    4   0.0   0.0    0.0   
6                   home_run       20          20   20  20.0  20.0    0.0   
7                     single       38          38   38  38.0   0.0    0.0   

   strikeouts   ba  obp  slg  ops   woba  
0         0.0  1.0  1.0  2.0  3.0  1.276  
1         0.0  0.0  0.0  0.0  0.0  0.000  
2         0.0  0.0  0.0  0.0  0.0  0.000  
3         0.0  0.0  0.0  0.0  0.0  0.000  
4         0.0  0.0  0.0  0.0  0.0  0.000  
5         0.0  0.0  0.0  0.0  0.0  0.000  
6         0.0  1.0  1.0  4.0  5.0  2.093  
7         0.0  1.0  1.0  1.0  2.0  0.894

In [45]:
results('game_year')

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2024      136         115  115    75   20    0.0         0.0  0.652   

     obp    slg    ops   woba  
0  0.652  1.322  1.974  0.848

In [46]:
df = pps[pps.game_year>=2020]
results('events').sort_values(by='pitches',ascending=False)
hr = df[df.events == 'home_run']
hr[(hr.game_year == 2024)&(hr.launch_speed == hr.groupby('game_year',as_index=False.agg({'launch_speed' : 'max'})).launch_speed.max())]

AttributeError: 'bool' object has no attribute 'agg'

In [ ]:
hr[(hr.game_year == 2024)&(hr.launch_speed == hr[hr.game_year == 2024].groupby('game_year',as_index=False).agg({'launch_speed' : 'max'}).launch_speed.max())]

In [ ]:
#df.groupby('home_team',as_index=False).agg({'des' : 'count','game_pk' : 'nunique'})
df = df[df.home_team.isin(teams)]
results('game_year')

In [47]:
df.hit_distance_sc.std() # This is too large to proceed

120.07232339618798

In [48]:
df[(df.home_team.isin(teams))
   &(df.game_year == 2024)
  ].groupby('home_team',as_index=False).agg({'game_pk':'nunique'})

NameError: name 'teams' is not defined

In [49]:
df = phils_sc
df[(df(df.home_team.isin(teams))&(df.launch_speed >= 106)&(df.launch_speed <= 107.2)
#&(df.launch_angle >=
&(df.hit_distance_sc >= hd-df.hit_distance_sc.std())&(df.hit_distance_sc <= hd+df.hit_idstance_sc.std()))]

NameError: name 'teams' is not defined

In [50]:
ls = 106.6
#la = 
hd = 413
available_dfs = [phils_sc,nphl]

teams = ['DET','MIN','TB','PHI','COL']
for df in available_dfs:
    df[(df(df.home_team.isin(teams))&(df.launch_speed >= 106)&(df.launch_speed <= 107.2)
    #&(df.launch_angle >=
    &(df.hit_distance_sc >= hd-df.hit_distance_sc.std())&(df.hit_distance_sc <= hd+df.hit_idstance_sc.std()))]


TypeError: 'DataFrame' object is not callable

In [51]:
pos.groupby('home_team',as_index=False).agg({'game_pk' : 'nunique'})

home_team  game_pk
0        ATL       79
1         AZ       29
2        BAL        4
3        BOS       14
4        CHC       27
5        CIN       31
6        CLE        6
7        COL       31
8        CWS        5
9        DET        5
10       HOU        9
11        KC        3
12       LAA        6
13       LAD       27
14       MIA       84
15       MIL       24
16       MIN        3
17       NYM       77
18       NYY       10
19       OAK        3
20       PHI      724
21       PIT       25
22        SD       30
23       SEA        5
24        SF       26
25       STL       34
26        TB       11
27       TEX        8
28       TOR       17
29       WSH       83

# Start Over - Kody Clemens Analysis
Kody Clemens is hitting the ball well in his increasing opportunities while Trea Turner recovers from his hamstring strain. Clemens was hitting well in AAA. As Rob Thomson says, he is a big league ballplayer. 

In [52]:
pos[pos.game_year == 2024].groupby('player_name',as_index=False).agg({'des' : 'count', 'batter' : 'max'
                                                                      ,'game_pk' : 'nunique'
                                                                     })
#kody_id = 


player_name  des  batter  game_pk
0          Bohm, Alec  786  664761       53
1   Castellanos, Nick  805  592206       54
2       Clemens, Kody  138  665019       15
3       Harper, Bryce  817  547180       49
4      Marsh, Brandon  750  669016       47
5    Merrifield, Whit  378  593160       26
6     Pache, Cristian  212  665506       17
7      Realmuto, J.T.  721  592663       43
8        Rojas, Johan  570  679032       44
9     Schwarber, Kyle  990  656941       52
10      Sosa, Edmundo  354  624641       25
11      Stott, Bryson  778  681082       49
12    Stubbs, Garrett  171  596117       14
13       Turner, Trea  576  607208       33

In [53]:
harper_id = 547180

In [54]:
harper = pd.concat([nphl[nphl.player_name == 'Harper, Bryce'], pos[pos.batter == harper_id]])
df = harper_phils = pos[pos.batter==harper_id]
ht = results('home_team')
ht.sort_values(by='ops',ascending=False)

home_team  pitches  plate_apps  bip  hits   hrs  walks  strikeouts     ba  \
25       TOR      118          29   17     7   4.0    7.0           5  0.318   
11       LAD      148          41   31    15   4.0    3.0           7  0.405   
20       SEA       59          14   10     6   0.0    2.0           1  0.500   
15       NYY       70          18    9     5   1.0    4.0           5  0.357   
26       WSH      711         176  119    50   8.0   23.0          32  0.331   
17       PHI     5664        1413  872   354  85.0  201.0         325  0.299   
23        TB      102          26   17     9   0.0    2.0           7  0.375   
4        CIN      103          26   18     6   2.0    4.0           4  0.273   
6        COL      254          69   46    18   5.0    5.0          18  0.281   
0        ATL      633         168  108    38  11.0   20.0          39  0.264   
18       PIT      183          42   19     9   0.0   10.0          13  0.281   
12       MIA      611         152   98    37   6.0   17.0          36  0.280   
24       TEX       18           4    2     1   0.0    1.0           1  0.333   
1         AZ      276          72   50    17   3.0    9.0          13  0.270   
3        CHC      208          63   50    19   1.0    4.0           9  0.328   
2        BOS      112          31   24     7   2.0    2.0           5  0.241   
16       OAK       40          10    8     4   0.0    0.0           2  0.400   
10       LAA       70          13    5     2   0.0    4.0           4  0.222   
5        CLE      107          24   15     5   1.0    3.0           6  0.250   
13       MIL      208          51   31     8   4.0    5.0          15  0.182   
21        SF      217          55   34    11   2.0    8.0          13  0.234   
19        SD      308          73   49    13   3.0    8.0          15  0.203   
9         KC       66          14    6     2   0.0    4.0           4  0.200   
14       NYM      473         114   69    20   2.0   12.0          31  0.202   
22       STL      221          55   32     7   2.0    7.0          15  0.149   
8        HOU       79          21   13     4   0.0    1.0           7  0.200   
7        DET       23           7    4     1   0.0    0.0           3  0.143   

      obp    slg    ops   woba  
25  0.483  0.909  1.392  0.546  
11  0.439  0.865  1.304  0.532  
20  0.571  0.667  1.238  0.531  
15  0.500  0.643  1.143  0.480  
26  0.426  0.570  0.996  0.424  
17  0.403  0.593  0.996  0.418  
23  0.423  0.542  0.965  0.413  
4   0.385  0.545  0.930  0.397  
6   0.333  0.594  0.927  0.387  
0   0.351  0.556  0.907  0.379  
18  0.452  0.438  0.890  0.395  
12  0.362  0.515  0.877  0.373  
24  0.500  0.333  0.833  0.393  
1   0.361  0.460  0.821  0.356  
3   0.365  0.448  0.813  0.350  
2   0.290  0.517  0.808  0.336  
16  0.400  0.400  0.800  0.353  
10  0.462  0.333  0.795  0.381  
5   0.333  0.450  0.783  0.328  
13  0.255  0.477  0.732  0.304  
21  0.345  0.383  0.728  0.322  
19  0.301  0.391  0.692  0.306  
9   0.429  0.200  0.629  0.321  
14  0.298  0.303  0.601  0.273  
22  0.273  0.319  0.592  0.267  
8   0.238  0.200  0.438  0.201  
7   0.143  0.286  0.429  0.174

In [55]:
player_id = pos[pos.player_name == 'Stott, Bryson'].batter.unique()[0]
df = pos[pos.batter == player_id]
df = df[df.game_year == 2024]
ht = results('home_team')
ht.sort_values(by='ops',ascending=False).reset_index(drop=True)

home_team  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       MIA       52          10    7     4  0.0    3.0         0.0  0.571   
1       NYM       48          10    7     3  1.0    3.0         0.0  0.500   
2       CIN       32           8    5     2  0.0    1.0         2.0  0.333   
3        SD       45          12    8     2  2.0    0.0         3.0  0.182   
4       LAA       39           7    5     1  0.0    2.0         0.0  0.200   
5       PHI      431         109   79    23  2.0   13.0        15.0  0.250   
6       WSH       37          10    7     2  0.0    2.0         1.0  0.250   
7       STL       52          13   10     2  0.0    0.0         3.0  0.167   
8       COL       42          13   11     1  0.0    1.0         1.0  0.083   

     obp    slg    ops   woba  
0  0.700  1.000  1.700  0.677  
1  0.600  1.000  1.600  0.597  
2  0.375  0.667  1.042  0.401  
3  0.250  0.727  0.977  0.410  
4  0.429  0.400  0.829  0.381  
5  0.349  0.348  0.696  0.318  
6  0.400  0.250  0.650  0.318  
7  0.154  0.167  0.321  0.138  
8  0.154  0.083  0.237  0.122

In [56]:
soto_pos = pps[pps.pitcher == 642397]
soto = pd.concat([nphl[nphl.player_name == 'Soto, Gregory'], soto_pos])
df = soto_pos
results('home_team')

home_team  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0        ATL       51          14    9   5.0  0.0    1.0         4.0  0.417   
1         AZ       31           8    5   2.0  0.0    2.0         1.0  0.333   
2        CHC       15           6    4   1.0  0.0    0.0         2.0  0.167   
3        CIN       49          12    7   2.0  0.0    1.0         4.0  0.200   
4        CLE       24           8    7   1.0  0.0    1.0         0.0  0.143   
5        COL       21           6    4   1.0  0.0    1.0         1.0  0.200   
6        CWS       22           5    2   0.0  0.0    0.0         3.0  0.000   
7        HOU       16           3    2   0.0  0.0    0.0         1.0  0.000   
8        LAA       32           9    6   2.0  0.0    1.0         2.0  0.250   
9        LAD       26           7    5   4.0  0.0    1.0         1.0  0.667   
10       MIA       66          20   16   9.0  1.0    1.0         3.0  0.474   
11       MIL       15           3    1   0.0  0.0    0.0         2.0  0.000   
12       NYM       34           9    6   1.0  0.0    0.0         3.0  0.111   
13       NYY       34           6    1   1.0  1.0    2.0         3.0  0.250   
14       OAK       14           2    2   0.0  0.0    0.0         0.0  0.000   
15       PHI      646         165   99  24.0  5.0   18.0        45.0  0.169   
16       PIT       23           7    5   2.0  0.0    0.0         2.0  0.286   
17        SD       18           5    4   2.0  0.0    1.0         0.0  0.667   
18        SF       31          11    8   5.0  0.0    0.0         3.0  0.455   
19       STL       45           9    6   3.0  0.0    1.0         2.0  0.375   
20        TB       20           4    3   1.0  0.0    0.0         1.0  0.250   
21       TEX       34           7    3   2.0  0.0    2.0         2.0  0.400   
22       TOR       11           3    2   0.0  0.0    0.0         1.0  0.000   
23       WSH       34          10   10   3.0  0.0    0.0         0.0  0.300   

      obp    slg    ops   woba  
0   0.429  0.583  1.012  0.417  
1   0.500  0.333  0.833  0.395  
2   0.167  0.167  0.333  0.147  
3   0.250  0.200  0.450  0.207  
4   0.250  0.143  0.393  0.197  
5   0.333  0.200  0.533  0.265  
6   0.000  0.000  0.000  0.000  
7   0.000  0.000  0.000  0.000  
8   0.333  0.250  0.583  0.276  
9   0.714  1.000  1.714  0.707  
10  0.500  0.789  1.289  0.546  
11  0.000  0.000  0.000  0.000  
12  0.111  0.111  0.222  0.099  
13  0.500  1.000  1.500  0.566  
14  0.000  0.000  0.000  0.000  
15  0.273  0.310  0.583  0.264  
16  0.286  0.286  0.571  0.252  
17  0.600  1.333  1.933  0.637  
18  0.455  0.455  0.909  0.401  
19  0.444  0.375  0.819  0.375  
20  0.250  0.250  0.500  0.221  
21  0.571  0.400  0.971  0.451  
22  0.000  0.000  0.000  0.000  
23  0.300  0.300  0.600  0.265

# Some Fun Puzzles Today

In [57]:
# Cristian Pache leads the NL in xBA? No.
# How many swinging strikes in a row?
# Wheeler does not handle adversity well
# Bohm pct of at_bats with 2 men on base

In [58]:
# I could put together the Wrike tasks that are completed in my name. Or the version history from tableau

In [59]:
bohm = pos[pos.player_name == 'Bohm, Alec']
df = bohm
results('home_team')
results('game_year')

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2020      661         180  126    54    4     16          36  0.338   
1       2021     1673         417  273    94    7     31         111  0.247   
2       2022     2518         696  529   177   14     33         126  0.274   
3       2023     2436         664  507   164   21     47         103  0.271   
4       2024      786         224  174    64    5     15          32  0.315   

     obp    slg    ops   woba  
0  0.400  0.481  0.881  0.381  
1  0.305  0.342  0.647  0.285  
2  0.310  0.393  0.704  0.306  
3  0.327  0.431  0.758  0.328  
4  0.366  0.502  0.869  0.378

In [60]:
gy = results('game_year')
ht = results('home_team')

In [61]:
gy['ppa'] = gy.pitches / gy.plate_apps
gy.round(3)

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2020      661         180  126    54    4     16          36  0.338   
1       2021     1673         417  273    94    7     31         111  0.247   
2       2022     2518         696  529   177   14     33         126  0.274   
3       2023     2436         664  507   164   21     47         103  0.271   
4       2024      786         224  174    64    5     15          32  0.315   

     obp    slg    ops   woba    ppa  
0  0.400  0.481  0.881  0.381  3.672  
1  0.305  0.342  0.647  0.285  4.012  
2  0.310  0.393  0.704  0.306  3.618  
3  0.327  0.431  0.758  0.328  3.669  
4  0.366  0.502  0.869  0.378  3.509

In [62]:
pos[pos.inning == 1][['on_1b','on_2b','on_3b']]

on_1b  on_2b  on_3b
99        NaN    NaN    NaN
102       NaN    NaN    NaN
104       NaN    NaN    NaN
111       NaN    NaN    NaN
113       NaN    NaN    NaN
...       ...    ...    ...
212047    NaN    NaN    NaN
212052    NaN    NaN    NaN
212055    NaN    NaN    NaN
212061    NaN    NaN    NaN
212066    NaN    NaN    NaN

[24976 rows x 3 columns]

In [63]:
x = df[ ############### All ################
  ((df.on_1b.isna()==True) & (df.on_2b.isna()==True) & (df.on_3b.isna() == True)) # Nobody On
  |
    ((df.on_1b.isna()==True) & (df.on_2b.isna()==True) & (df.on_3b.isna() == False))
  |
    ((df.on_1b.isna()==True) & (df.on_2b.isna()==False) & (df.on_3b.isna() == True)) # Runner on Second
  |
    ((df.on_1b.isna()==False) & (df.on_2b.isna()==True) & (df.on_3b.isna() == True)) # Runner on First
  |
    ((df.on_1b.isna()==True) & (df.on_2b.isna()==False) & (df.on_3b.isna() == False)) # Runner on Second + Third
  |
    ((df.on_1b.isna()==False) & (df.on_2b.isna()==True) & (df.on_3b.isna() == False)) # Runner on First + Third
  |
    ((df.on_1b.isna()==False) & (df.on_2b.isna()==False) & (df.on_3b.isna() == True)) # Runner on First + Second
  |
    ((df.on_1b.isna()==False) & (df.on_2b.isna()==False) & (df.on_3b.isna() == False)) # Basees Loaded
  ]
bohm = results('game_year')
bohm

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2020      661         180  126    54    4     16          36  0.338   
1       2021     1673         417  273    94    7     31         111  0.247   
2       2022     2518         696  529   177   14     33         126  0.274   
3       2023     2436         664  507   164   21     47         103  0.271   
4       2024      786         224  174    64    5     15          32  0.315   

     obp    slg    ops   woba  
0  0.400  0.481  0.881  0.381  
1  0.305  0.342  0.647  0.285  
2  0.310  0.393  0.704  0.306  
3  0.327  0.431  0.758  0.328  
4  0.366  0.502  0.869  0.378

In [64]:
# Define a Quality At Bat

In [65]:
df[((df.on_1b.isna() == True) &(df.on_2b.isna() == False) &(df.on_3b.isna() == False))
   #|((df.on_1b.isna() == False) &(df.on_2b.isna() == True) &(df.on_3b.isna() == False))
  ]['on_1b'].unique()

array([nan])

In [66]:
df[['on_1b','on_2b','on_3b']].head(25)

on_1b     on_2b  on_3b
119199       NaN       NaN    NaN
119200       NaN       NaN    NaN
119201       NaN       NaN    NaN
119233       NaN       NaN    NaN
119234       NaN       NaN    NaN
119235       NaN       NaN    NaN
119236       NaN       NaN    NaN
119237       NaN       NaN    NaN
119264       NaN       NaN    NaN
119270       NaN       NaN    NaN
119273       NaN       NaN    NaN
119274       NaN       NaN    NaN
119276       NaN       NaN    NaN
119277       NaN       NaN    NaN
119280       NaN       NaN    NaN
119281       NaN       NaN    NaN
119315       NaN       NaN    NaN
119316       NaN       NaN    NaN
119364       NaN  457705.0    NaN
119368       NaN  457705.0    NaN
119382  457705.0       NaN    NaN
119384  457705.0       NaN    NaN
119391  457705.0       NaN    NaN
119397  457705.0       NaN    NaN
119402  457705.0       NaN    NaN

In [67]:
df.columns.tolist()

['pitch_type',
 'game_date',
 'release_speed',
 'release_pos_x',
 'release_pos_z',
 'player_name',
 'batter',
 'pitcher',
 'events',
 'description',
 'spin_dir',
 'spin_rate_deprecated',
 'break_angle_deprecated',
 'break_length_deprecated',
 'zone',
 'des',
 'game_type',
 'stand',
 'p_throws',
 'home_team',
 'away_team',
 'type',
 'hit_location',
 'bb_type',
 'balls',
 'strikes',
 'game_year',
 'pfx_x',
 'pfx_z',
 'plate_x',
 'plate_z',
 'on_3b',
 'on_2b',
 'on_1b',
 'outs_when_up',
 'inning',
 'inning_topbot',
 'hc_x',
 'hc_y',
 'tfs_deprecated',
 'tfs_zulu_deprecated',
 'fielder_2',
 'umpire',
 'sv_id',
 'vx0',
 'vy0',
 'vz0',
 'ax',
 'ay',
 'az',
 'sz_top',
 'sz_bot',
 'hit_distance_sc',
 'launch_speed',
 'launch_angle',
 'effective_speed',
 'release_spin_rate',
 'release_extension',
 'game_pk',
 'pitcher.1',
 'fielder_2.1',
 'fielder_3',
 'fielder_4',
 'fielder_5',
 'fielder_6',
 'fielder_7',
 'fielder_8',
 'fielder_9',
 'release_pos_y',
 'estimated_ba_using_speedangle',
 'estimat

df['men_on_base'] = df.transform('

Quality At Bats:
    - H, BB, HBP, SH, SF, Well-hit out, df.pitch_number.max() >= '8'

In [68]:
df['ab'] = df.game_pk + df.at_bat_number
df

pitch_type   game_date  release_speed  release_pos_x  release_pos_z  \
119199         SI  2020-09-27           91.0           2.53           5.73   
119200         FC  2020-09-27           85.6           2.52           5.71   
119201         SI  2020-09-27           90.9           2.53           5.71   
119233         SI  2020-09-27           92.0           2.45           5.75   
119234         SI  2020-09-27           91.0           2.61           5.77   
...           ...         ...            ...            ...            ...   
211961         FF  2024-03-29           96.1          -1.74           6.00   
211996         FF  2024-03-29           97.5          -1.77           5.95   
212034         FF  2024-03-29           95.0          -2.19           5.85   
212035         FC  2024-03-29           87.7           3.06           5.29   
212041         FF  2024-03-29           94.3          -2.22           5.94   

       player_name  batter  pitcher     events    description  ...    w1B  \
119199  Bohm, Alec  664761   676596  field_out  hit_into_play  ...  0.883   
119200  Bohm, Alec  664761   676596        NaN           ball  ...  0.883   
119201  Bohm, Alec  664761   676596        NaN           ball  ...  0.883   
119233  Bohm, Alec  664761   676596  field_out  hit_into_play  ...  0.883   
119234  Bohm, Alec  664761   676596        NaN  called_strike  ...  0.883   
...            ...     ...      ...        ...            ...  ...    ...   
211961  Bohm, Alec  664761   675911        NaN           foul  ...  0.894   
211996  Bohm, Alec  664761   675911     single  hit_into_play  ...  0.894   
212034  Bohm, Alec  664761   641729     single  hit_into_play  ...  0.894   
212035  Bohm, Alec  664761   607481  force_out  hit_into_play  ...  0.894   
212041  Bohm, Alec  664761   641729        NaN  called_strike  ...  0.894   

          w2B    w3B    wHR  runSB  runCS   R/PA     R/W   cFIP      ab  
119199  1.238  1.558  1.979    0.2 -0.435  0.125  10.282  3.191  630989  
119200  1.238  1.558  1.979    0.2 -0.435  0.125  10.282  3.191  630989  
119201  1.238  1.558  1.979    0.2 -0.435  0.125  10.282  3.191  630989  
119233  1.238  1.558  1.979    0.2 -0.435  0.125  10.282  3.191  630972  
119234  1.238  1.558  1.979    0.2 -0.435  0.125  10.282  3.191  630972  
...       ...    ...    ...    ...    ...    ...     ...    ...     ...  
211961  1.276  1.619  2.093    0.2 -0.402  0.116   9.613  3.158  745632  
211996  1.276  1.619  2.093    0.2 -0.402  0.116   9.613  3.158  745617  
212034  1.276  1.619  2.093    0.2 -0.402  0.116   9.613  3.158  745650  
212035  1.276  1.619  2.093    0.2 -0.402  0.116   9.613  3.158  745681  
212041  1.276  1.619  2.093    0.2 -0.402  0.116   9.613  3.158  745650  

[8074 rows x 109 columns]

In [69]:
df[['game_pk','at_bat_number','ab']]

game_pk  at_bat_number      ab
119199   630945             44  630989
119200   630945             44  630989
119201   630945             44  630989
119233   630945             27  630972
119234   630945             27  630972
...         ...            ...     ...
211961   745604             28  745632
211996   745604             13  745617
212034   745604             46  745650
212035   745604             77  745681
212041   745604             46  745650

[8074 rows x 3 columns]

In [70]:
x = df.groupby(['game_pk','at_bat_number'],as_index=False
          ).agg({'pitch_number' : 'max'
                 ,'pitch_type' : 'nunique'
                 ,'des' : 'max'
                 ,'events' : 'unique'
                 #,'game_pk' : 'count'
                })

In [71]:
x['event'] = ''
for i in range(len(x)):
    x['event'].iloc[i] = x['events'].iloc[i][0]
    #x.groupby(x['events'].iloc[0][0]
    
x

game_pk  at_bat_number  pitch_number  pitch_type  \
0      630273             11             3           2   
1      630273             31             5           2   
2      630273             43             4           2   
3      630273             61             2           1   
4      630324             12             5           2   
...       ...            ...           ...         ...   
2179   748575             53             2           1   
2180   748577              6             2           2   
2181   748577             21             3           2   
2182   748577             37             7           3   
2183   748577             55             5           3   

                                                    des            events  \
0     Alec Bohm homers (1) on a fly ball to center f...   [home_run, nan]   
1     Alec Bohm singles on a ground ball to right fi...     [single, nan]   
2                                      Alec Bohm walks.       [walk, nan]   
3     Alec Bohm singles on a line drive to right fie...     [single, nan]   
4     Alec Bohm grounds out sharply, shortstop Dansb...  [field_out, nan]   
...                                                 ...               ...   
2179  Alec Bohm grounds out, shortstop Jon Berti to ...  [field_out, nan]   
2180  Alec Bohm lines out to right fielder Jesus San...  [field_out, nan]   
2181  Alec Bohm doubles (1) on a line drive to left ...     [double, nan]   
2182     Alec Bohm pops out to first baseman Josh Bell.  [field_out, nan]   
2183                    Alec Bohm strikes out swinging.  [strikeout, nan]   

          event  
0      home_run  
1        single  
2          walk  
3        single  
4     field_out  
...         ...  
2179  field_out  
2180  field_out  
2181     double  
2182  field_out  
2183  strikeout  

[2184 rows x 7 columns]

# That is Just Not True!
Ranger Suarez has been incredible this season. He starts tomorrow against the Gigantes.

In [76]:
ranger = pps.groupby(['p_throws','player_name'],as_index=False).agg({'pitcher' : 'max'
                                                                     ,'des' : 'count'
                                                                     ,'game_pk' : 'nunique'
                                                                    })
ranger[ranger.p_throws == 'L'].sort_values(by='des',ascending=False)
ranger_id = 624133
ranger_pps = pps[pps.pitcher == ranger_id]
ranger24 = ranger_pps[ranger_pps.game_year == 2924]
df = ranger_pps
results('game_year')

game_year  pitches  plate_apps  bip  hits  hrs  walks  strikeouts     ba  \
0       2018      225          68   52    21    3      5          11  0.339   
1       2019      763         205  150    52    6     10          42  0.275   
2       2020       94          26   20    10    1      4           1  0.476   
3       2021     1614         415  273    73    4     30         107  0.194   
4       2022     2807         725  514   158   16     64         142  0.242   
5       2023     2289         612  421   142   14     50         138  0.257   
6       2024     1040         271  175    43    5     16          77  0.171   

     obp    slg    ops   woba  
0  0.382  0.548  0.931  0.394  
1  0.307  0.413  0.720  0.303  
2  0.577  0.714  1.291  0.545  
3  0.260  0.255  0.515  0.234  
4  0.310  0.382  0.693  0.305  
5  0.317  0.395  0.712  0.310  
6  0.229  0.267  0.496  0.226